runtime/src/read_optimized_dashmap.rs

```diff
@@ -0,0 +1,199 @@
+#![allow(dead_code)]
+
+#[cfg(feature = "shuttle-test")]
+use shuttle::sync::Arc;
+#[cfg(not(feature = "shuttle-test"))]
+use std::sync::Arc;
+use {
+    dashmap::{
+        mapref::{entry::Entry, multiple::RefMulti},
+        DashMap,
+    },
+    std::{hash::BuildHasher, ops::Deref},
+};
+
+type DashmapIteratorItem<'a, K, V, S> = RefMulti<'a, K, ROValue<V>, S>;
+
+// Wrapper around dashmap that stores (K, Arc<V>) to minimize shard contention.
+#[derive(Debug)]
+pub struct ReadOptimizedDashMap<K, V, S>
+where
+    K: Clone + Eq + std::hash::Hash,
+    S: Clone + BuildHasher,
+{
+    inner: DashMap<K, ROValue<V>, S>,
+}
+
+impl<K, V, S> ReadOptimizedDashMap<K, V, S>
+where
+    K: Clone + Eq + std::hash::Hash,
+    S: Clone + BuildHasher,
+{
+    pub fn new(inner: DashMap<K, ROValue<V>, S>) -> Self {
+        Self { inner }
+    }
+
+    /// Alternative to entry(k).or_insert_with(default) that returns an Arc<V> instead of returning a
+    /// guard that holds the underlying shard's write lock.
+    ///
+    /// # Safety
+    ///
+    /// - Care must be taken when mutating the returned value. If modifications are applied while
+    ///   another thread concurrently deletes the key, the changes may be lost.
+    pub unsafe fn get_or_insert_with(&self, k: &K, default: impl FnOnce() -> V) -> ROValue<V> {
+        match self.inner.get(k) {
+            Some(v) => ROValue::clone(&*v),
+            None => ROValue::clone(
+                self.inner
+                    .entry(k.clone())
+                    .or_insert_with(|| ROValue::new(default()))
+                    .value(),
+            ),
+        }
+    }
+
+    /// Returns an Arc clone of the value corresponding to the key.
+    ///
+    /// # Safety
+    ///
+    /// - Care must be taken when mutating the returned value. If modifications are applied while
+    ///   another thread concurrently deletes the key, the changes may be lost.
+    pub unsafe fn get(&self, k: &K) -> Option<ROValue<V>> {
+        self.inner.get(k).map(|v| ROValue::clone(&v))
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = DashmapIteratorItem<K, V, S>> {
+        self.inner.iter()
+    }
```

Comment:
nit - may be useful to propagate the lock warning here from dashmap docs? If you're holding a write-lock on map and call this it may deadlock.

```suggestion
    /// See [`DashMap::iter`] for details about locking semantics and guarantees.
    pub fn iter(&self) -> impl Iterator<Item = DashmapIteratorItem<K, V, S>> {
        self.inner.iter()
    }
```